IM_calculation/source_site_dist/src_site_dist.py

```diff
@@ -114,27 +144,74 @@ def calc_rx_ry(srf_points: np.ndarray, plane_infos: List[Dict], locations: np.nd
                 down_strike_top_point,
                 up_strike_top_point,
             )
-            up_strike_bottom_point, down_strike_bottom_point = (
-                down_strike_bottom_point,
-                up_strike_bottom_point,
-            )
 
-        r_x[iloc] = ll_cross_track_dist(
+        r_x[iloc], r_y[iloc] = ll_cross_along_track_dist(
             *up_strike_top_point, *down_strike_top_point, lon, lat
         )
 
-        up_strike_dist = ll_cross_track_dist(
-            *up_strike_top_point, *up_strike_bottom_point, lon, lat
-        )
-        down_strike_dist = ll_cross_track_dist(
-            *down_strike_top_point, *down_strike_bottom_point, lon, lat
-        )
+    return r_x, r_y
 
-        if np.sign(up_strike_dist) != np.sign(down_strike_dist):
-            # If the signs are different then the point is between the lines projected along the edges of the srf
-            r_y[iloc] = 0
-        else:
-            r_y[iloc] = np.min(np.abs([up_strike_dist, down_strike_dist]))
+
+def calc_rx_ry_GC2(
+    srf_points: np.ndarray,
+    plane_infos: List[Dict],
+    locations: np.ndarray,
+    hypocentre_origin=False,
+):
+    """
+    Calculates Rx and Ry distances using the cross track and along track distance calculations
+    If there are multiple fault planes the Rx, Ry values are calculated for each fault plane individually, then weighted
+    according to plane length and distance to the location
+    For one fault plane this is the same as the GC1 function
+    :param srf_points: An array with shape (n, 3) giving the lon, lat, depth location of each subfault
+    :param plane_infos: A list of srf header dictionaries, as retrieved from qcore.srf.get_headers with idx=True
+    :param locations: An array with shape (m, 2) giving the lon, lat locations of each location to get Rx, Ry values for
+    :param hypocentre_origin: If True sets the Ry origin/0 point to the fault trace projection of the hypocentre. If
+    false the most upstrike subfault of the first fault trace is used
+    :return: An array with shape (m, 2) giving the Rx, Ry values for each of the given locations
+    """
+    r_x = np.empty(locations.shape[0])
+    r_y = np.empty(locations.shape[0])
+
+    # Separate the srf points into the different plane traces
+    pnt_counts = [plane["nstrike"] * plane["ndip"] for plane in plane_infos]
+    pnt_counts.insert(0, 0)
+    pnt_counts = np.cumsum(pnt_counts)
+    pnt_sections = [
+        srf_points[pnt_counts[i] : pnt_counts[i] + header["nstrike"]]
+        for i, header in enumerate(plane_infos)
+    ]
+
+    origin_offset = 0
+    if hypocentre_origin:
+        # Our faults only use one hypocentre
+        # Will only use the first one found if there are multiple
+        for plane in plane_infos:
+            if plane["shyp"] == -999.9000:
+                origin_offset -= plane["length"]
+            else:
+                origin_offset -= plane["length"] / 2 + plane["shyp"]
+                break
```

Comment:
This isn't my favourite block of logic. But I can't think of a better way to to it.

Reply:
It would be perfect for a do while loop :P